### PR TITLE
TwoLocal entangling map works for numpy integers

### LINKED ISSUE #6455: TwoLocal entangling map does not work for numpy integers
### Informations

- **Qiskit version**: 0.17.1
- **Python version**: 3
- **Operating system**: ubuntu 18

### What is the current behavior?
When generating an entangling map by virtue of numpy arrays (e.g. np.arange(num)) it raises a ValueError 
```ValueError: Invalid value of entanglement: [[(0, 1), (1, 2), (2, 3), (3, 4), (4, 5)]]```

### Steps to reproduce the problem
Minimal working example

```
import qiskit
import numpy as np

L = 6

def create_map(L):
    return [(i,i+1) for i in np.arange(L-1)]

map1 = create_map(L)
print(map1)

map_by_hand1 = [(0, 1), (1, 2), (2, 3), (3, 4), (4, 5)]
print(map_by_hand1)

print(map1 == map_by_hand1) # returns True

ansatz = qiskit.circuit.library.TwoLocal(L,"ry","cz",[map_by_hand1],reps=2, insert_barriers=True)
ansatz.draw("mpl") # works

ansatz = qiskit.circuit.library.TwoLocal(L,"ry","cz",[map1],reps=2, insert_barriers=True)
ansatz.draw("mpl") # raises ValueError
```

### What is the expected behavior?
Since map1 == map_by_hand1 is the same value, it is expected that both are viable options to pass as entangling maps.

### Suggested solutions
change the isinstance() check in n_local

### PR DESCRIPTION
**Pull request by ishworii**
### Summary

Fixes #6455 

### Details and comments
Previously TwoLocal entangling map did not accept numpy integers as input
```
from qiskit.circuit.library import TwoLocal
import numpy as np

num_qubits = 5

map_np = [(i,i+1) for i in np.arange(num_qubits - 1)]

two = TwoLocal(num_qubits,'ry','cx',entanglement = map_np)
two.draw('mpl')
```

This would raise ValueError.
In this PR, the above issue has been fixed by adding the proper instance check.

### CODE REVIEW

**Comment by ishworii** [comment]
So this PR fixes the issue where TwoLocal doesn't accept numpy integers but while working I discovered that even though it works for TwoLocal, when we compose the circuit it doesn't work.
For example
```
from qiskit.circuit import QuantumCircuit
from qiskit.circuit.library import TwoLocal
import numpy as np

expected = [(i, i + 1) for i in np.arange(num_qubits - 1, dtype=np.int64)]

qc = QuantumCircuit(num_qubits)
two = TwoLocal(num_qubits,'ry','cx',entanglement = expected)
qc.compose(two,inplace = True)
```

This generates KeyError,possibly due to lack of instance check for np.int32 or np.int64 in 

https://github.com/Qiskit/qiskit-terra/blob/72567d5f8ccad4f7ccbb39f6a6b8645ff55c99f9/qiskit/circuit/quantumcircuit.py#L747-L762

Do I have to create new issue and PR to resolve that or can I update it here?

**Comment by Cryoris** [comment]
I think this should be addressed in this PR as well. You can probably just cast the numpy ints to Python `int`s instead of adding them in the instance check 🙂 Also, you can use `np.integer` as reference type which covers both `np.int32` and `np.int64`.

**Comment by ishworii** [comment]
> You can probably just cast the numpy ints to Python `int`s instead of adding them in the instance check

Should I do this 

https://github.com/Qiskit/qiskit-terra/blob/72567d5f8ccad4f7ccbb39f6a6b8645ff55c99f9/qiskit/circuit/library/n_local/n_local.py#L597-L630


here too?

**Comment by ishworii** [comment]
Also, if we do casting in n_local.py , we don't have to worry about follow up issue. @Cryoris

**Comment by ishworii** [comment]
@Cryoris Can you please review it?

File: test/python/circuit/library/test_nlocal.py [hunk 1]
```diff
@@ -732,6 +732,18 @@ def test_circular_on_same_block_and_circuit_size(self):
...
+        """Test if TwoLocal can be made and drawn from numpy integers"""
+        num_qubits = 6
+        expected_np32 = [(i, i + 1) for i in np.arange(num_qubits - 1, dtype=np.int32)]
+        expected_np64 = [(i, i + 1) for i in np.arange(num_qubits - 1, dtype=np.int64)]
+
+        two_np32 = TwoLocal(num_qubits, "ry", "cx", entanglement=expected_np32)
+        two_np64 = TwoLocal(num_qubits, "ry", "cx", entanglement=expected_np64)
+
+        self.assertTrue(two_np32.draw())
+        self.assertTrue(two_np64.draw())
```
Comment (Cryoris) [inline test/python/circuit/library/test_nlocal.py:745]: Could you exchange this for a more quantitative test? E.g. counting the number of entanglement gates
```python
expected_cx = 3 * num_qubits * (num_qubits - 1) / 2  # full entanglement and 3 reps
self.assertTrue(two_np32.count_ops()['cx'], expected_cx)
```

**Comment by ishworii** [comment]
@Cryoris Does this work?

**Review by Cryoris — APPROVED** [review]
LGTM thanks for the update!

**Comment by Cryoris** [comment]
Actually, could you add a bugfix releasenote saying that numpy integers are now also supported?

File: releasenotes/notes/bugfix-supports-numpy-integers-03c60659fe4cc0c5.yaml [hunk 2]
```diff
@@ -0,0 +1,7 @@
+---
+
+fixes:
+  - |
+    Fixes an issue where :class:`~qiskit.circuit.library.TwoLocal' did not
```
Comment (Cryoris) [inline releasenotes/notes/bugfix-supports-numpy-integers-03c60659fe4cc0c5.yaml:5]:
```suggestion
    Fixes an issue where :class:`~qiskit.circuit.library.TwoLocal` did not
```

**Comment by ishworii** [comment]
Done! @Cryoris

File: releasenotes/notes/bugfix-supports-numpy-integers-03c60659fe4cc0c5.yaml [hunk 3]
```diff
@@ -0,0 +1,7 @@
...
+    support numpy integers as input for ``entangling map ``. This has
```
Comment (Cryoris) [inline releasenotes/notes/bugfix-supports-numpy-integers-03c60659fe4cc0c5.yaml:6]:
```suggestion
    support numpy integers as input for ``entanglement``. This has
```

**Review by Cryoris — APPROVED** [review]
LGTM, thanks for the fix!

**Comment by ishworii** [comment]
@Cryoris I couldn't understand the check failing here, can you tell me what should I do?

**Pull request merged**